WIP: Rework exercises (#8)
sfeir-open-source/sfeir-school-rxjs: exercises/react-chat/README.md

# React Chat

The purpose of this exercise is to build a functional chat application with React and RxJS.

## Start the application

### Start the application with a local server
To start the application with a local server, run `npm run localstart`  
Keep in mind you will only be able to chat with yourself !

### Start the application and join an existing server
Replace the `SOCKET` url in `constants.js` with the url of the server and run `npm start`

### Start a public server
Run `npm run serve` and `npm run ngrok` and give the generated URL to the participants.

## First step

### Log in 
1. Create the `username$` observable using the constructor `new Observable(...)` that listens to the event `new-user` sent by the server.

2. In the function `susbscribeToUsername`, subscribe to the observable `username$` and set the username accordingly. The server send an error if the username is already taken, deal with the error case !

#### Tips

> * Use `socket.on('event-name')` to listens to an event sent by the server

Reviewer: ```suggestion
* Use `socket.on('event-name')` to listen to an event sent by the server
```

> * The server's reponse looks like this : `{ ok: boolean, username: "string" }` with `ok` being false when a username is already taken.

Reviewer: je ne comprends pas pourquoi boolean est écrit comme un type mais string comme une valeur, donc je fais cette suggestion, mais je vois que string est écrit comme une valeur à d’autres endroits, donc si j’ai vu juste il faudra changer partout.
```suggestion
* The server's response looks like this : `{ ok: boolean, username: string }` with `ok` being false when a username is already taken.
```

* Use the observer's error callback to deal with the error case

### Display online users
1. Create the `users$` observable using the constructor `new Observable(...)` that listens to the event `refresh-users` sent by the server.

> 2. In the function `susbscribeToUsers`, subscribe to the observable `users$` ans set the users list accordingly.

Reviewer: ```suggestion
2. In the function `subscribeToUsers`, subscribe to the observable `users$` and set the user list accordingly.
```

#### Tips
* The server's reponse looks like this : `["user1", "user2"]`

### Display messages
1. Create the `messages$` observable using the constructor `new Observable(...)` that listens to the event `new-message` sent by the server.

> 2. In the function `susbscribeToMessages`, subscribe to the observable `messages$` and add the new message to the message list.

Reviewer: ```suggestion
2. In the function `subscribeToMessages`, subscribe to the observable `messages$` and add the new message to the message list.
```
